telemetry/class-telemetry-event-queue.php

```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * Telemetry: event queue.
+ *
+ * @package Automattic\VIP\Telemetry
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\VIP\Telemetry;
+
+use WP_Error;
+
+/**
+ * Handles queued events to be sent to a telemetry service.
+ */
+class Telemetry_Event_Queue {
+
+	/**
+	 * @var Telemetry_Client The client to use to record events.
+	 */
+	private Telemetry_Client $client;
+
+	/**
+	 * Events queued to be sent to the telemetry service.
+	 *
+	 * @var array<Telemetry_Event>
+	 */
+	protected array $events = array();
+
+	/**
+	 * Constructor. Registers the shutdown hook to record any and all events.
+	 */
+	public function __construct( Telemetry_Client $client ) {
+		$this->client = $client;
+
+		// Register the shutdown hook to record any and all events
+		add_action( 'shutdown', array( $this, 'record_events' ) );
+	}
+
+	/**
+	 * Enqueues an event to be recorded asynchronously.
+	 *
+	 * @param Telemetry_Event $event The event to record.
+	 * @return bool|WP_Error True if the event was enqueued for recording.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function record_event_asynchronously( Telemetry_Event $event ): bool|WP_Error {
```

Comment:
Intersection types are a PHP 8.1 feature; do we still have sites running PHP 8.0?

Reply:
It is 8.0 feature: https://3v4l.org/rgSko#v8.0.30